Remove Vector<T> fallbacks for Vector128<T>

stephentoub opened the pull request:
I believe the only reason we were keeping these around were some mono platforms that accelerated `Vector<T>` but not `Vector128<T>`...

@fanyang-mono, @radekdoulik, are we able to remove these now?

stephentoub committed:
Remove Vector<T> fallbacks for Vector128<T>

adamsitnik reviewed (commented):
This PR warms my hart, I hope we can remove this code path ;)

stephentoub commented:
@SamMonoRT, any insights here?

fanyang-mono commented:
In .NET8, we worked on intrinsifying Vector128 for mini JIT on arm64, but we haven't do that for amd64. I would hold on doing this, until that work is finished.

stephentoub commented:
Thanks.  Do we have an ETA for appx when that will happen?

radekdoulik commented:
On wasm this should be fine, the Vector128<T> coverage is pretty good. I can fill in if anything would regress.

SamMonoRT commented:
> In .NET8, we worked on intrinsifying Vector128 for mini JIT on arm64, but we haven't do that for amd64. I would hold on doing this, until that work is finished.

@fanyang-mono  - We implemented a few Vector128 APIs on AMDx64 via our intern last year. Do we have a list of remaining ones ? 

@stephentoub  - AMDx64 is low priority for us, but we will evaluate and consider implementing remaining ones (based on Fan's response) this summer. I wouldn't expect that to be completed prior to Preview 7 at this time.

fanyang-mono commented:
The intern added support for Vector128 support with LLVM as backend for AMD64 last summer. We used the doc project file to track at that time. I will create an issue to clarify this, as what I remembered that most of the API's should be supported.

fanyang-mono commented:
> Thanks. Do we have an ETA for appx when that will happen?

I don't know atm, maybe in .NET9 time frame, but can't say for sure.

stephentoub commented:
@fanyang-mono, are you saying then that it's important these `Vector<T>` code paths remain here for .NET 8? 😦

fanyang-mono commented:
> @fanyang-mono, are you saying then that it's important these `Vector<T>` code paths remain here for .NET 8? 😦

I just discussed this topic with my manager Sam, we understood the importance of this code change and are currently evaluating our options. Will keep you posted. The work is tracked by https://github.com/dotnet/runtime/issues/86272

stephentoub commented:
To be clear, we've lived with it for this long, we can live with it for a bit longer if necessary.  It'd just be really nice to get rid of :-)

stephentoub commented:
To at least make some progress here, I will ifdef these code paths to just be for mono for now rather than deleting them entirely.  Then at least for coreclr and nativeaot it won't impact corelib binaries for .NET 8.

stephentoub committed:
ifdef instead of deleting Vector<T> code paths

ifdef them to mono as they're not used by coreclr or nativeaot.  They can be deleted instead once mono's Vector128 support improves.

stephentoub commented:
I've updated the PR to just ifdef the `Vector<T>` paths to be mono-specific instead of outright deleting them.
cc: @tannergooding

stephentoub commented:
@SamMonoRT, this should be a nop for mono so I assume you're ok with it. We can delete the code paths and improve maintainability when mono is ready

@adamsitnik, @egorbo, or @tannergooding, mind reviewing?

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:
```diff
@@ -1069,6 +1079,22 @@ private static int LocateLastFoundByte(Vector<byte> match)
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector<byte> LoadVector(ref byte start, nuint offset)
```
nit: this helper is not needed because there is `Vector.LoadUnsafe`

stephentoub replied:
I've not changed any code.  I only moved a few methods in the source to reduce the number of ifdefs.

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:
```diff
@@ -1044,6 +1032,28 @@ public static nuint CommonPrefixLength(ref byte first, ref byte second, nuint le
+#if MONO
+        // Vector sub-search adapted from https://github.com/aspnet/KestrelHttpServer/pull/1138
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int LocateFirstFoundByte(Vector<byte> match)
```
this looks to be a sub-optimal implementation (not using the MoveMask/TLZ trick like we do for V128/V256) but since we plan to get rid of it I guess it doesn't matter

stephentoub replied:
I've not changed any code. I only moved a few methods in the source to reduce the number of ifdefs.

SamMonoRT commented:
@fanyang-mono - need to make sure we track the cleanup of this file once the implementations are completed on our side.

fanyang-mono commented:
> @fanyang-mono - need to make sure we track the cleanup of this file once the implementations are completed on our side.

Updated #86272

stephentoub commented:
Thanks, all
